xcdat/bounds.py

```diff
@@ -245,33 +257,80 @@ def add_bounds(self, axis: CFAxisName, width: float = 0.5) -> xr.Dataset:
             If bounds already exist. They must be dropped first.
 
         """
+        ds = self._dataset.copy()
         self._validate_axis_arg(axis)
 
+        coord_vars: Union[xr.DataArray, xr.Dataset] = get_axis_coord(
+            self._dataset, axis
+        )
+
+        for coord in coord_vars.coords.values():
+            # Check if the coord var has a "bounds" attr and the bounds actually
+            # exist in the Dataset. If it does not, then add the bounds.
+            try:
+                bounds_key = ds[coord.name].attrs["bounds"]
+                ds[bounds_key]
+
+                continue
+            except KeyError:
+                bounds = self._create_bounds(axis, coord, width)
+
+                ds[bounds.name] = bounds
+                ds[coord.name].attrs["bounds"] = bounds.name
```

Comment:
Continuing on this thread, I hackily fixed this by updating `_create_bounds` [here](https://github.com/xCDAT/xcdat/pull/343/files#diff-6a74d16440d82ae43ed458dc26fafd0427c8568b34915e698fce57b2c9a8a7e1R351-R356) to:

```
        if len(coord_var.shape) < 2:
            return None
        elif coord_var.shape[0] <= 1:
            raise ValueError(
                f"Cannot generate bounds for coordinate variable '{coord_var.name}'"
                " which has a length <= 1."
            )
```

I assume there is a better way to deal with this. It was having issues because the shape of `height` was `()` (it is a float in a dataarray. Also: should we really raise an error if this issue comes up? I thought we could just return `None` in both cases and only add the bounds if they can be generated (this would require adjusting the documentation) and would mean changing the above code to:

```
                bounds = self._create_bounds(axis, coord, width)
                if bounds:
                    ds[bounds.name] = bounds
                    ds[coord.name].attrs["bounds"] = bounds.name
```

One other question. When I add these changes, I still get this Warning: 

> 2022-10-11 19:19:34,748 [WARNING]: dataset.py(decode_time:333) >> height does not have a 'calendar' attribute set, so it could not be decoded. Set the 'calendar' attribute (`ds.height.attrs['calendar]`) and try decoding them again

I don't understand why height is getting picked up in decode_time (yet).

Reply:
As you mentioned, `height` is a singleton with a type `float64`.
I noticed `height` is not a dimension and it is included for all dimensions as a non-dimensional coordinate.
I'm trying to figure out if this is the default xarray behavior and if so why it is.

```python
import xarray as xr

fn = "/p/css03/esgf_publish/CMIP6/CMIP/CSIRO-ARCCSS/ACCESS-CM2/historical/r1i1p1f1/Amon/tas/gn/v20191108/tas_Amon_ACCESS-CM2_historical_r1i1p1f1_gn_185001-201412.nc"
ds_xr = xr.open_dataset(fn)

print(ds.time.coords)
"""
Coordinates:
  * time     (time) float64 15.5 45.0 74.5 ... 6.019e+04 6.022e+04 6.025e+04
    height   float64 2.0
"""

print(ds_xr.lat.coords)
"""
Coordinates:
  * lat      (lat) float64 -89.38 -88.12 -86.88 -85.62 ... 86.88 88.12 89.38
    height   float64 2.0
"""
```

For `decode_times()`, it loops over time dimension coordinates (which includes `height`) and tries to decode them.
```
from xcdat import get_axis_coord()

coord_keys = list(get_axis_coord(ds, "T").coords)

print(coord_keys)
# ['time', 'height']
```

Reply:
This xarray GH issue seems to be related: [[Bug]: Wrong list of coordinate when a singleton coordinate exists #6196  ](https://github.com/pydata/xarray/issues/6196)

Reply:
I think when I was trying to solve some of these issues, I checked for keys that were both a `dim` and a `coord` for a given axis/dataarray when retrieving them by CF label (e.g., `X`, `Y`, and `T`). Note that 

`ds.time.dims` returns:
> ('time',)

I'm not sure if this could be a useful constraint (or maybe it will cause other problems?).

Reply:
To follow-up on this...I suggest an alternate solution (to the code I wrote above) here: https://github.com/xCDAT/xcdat/pull/343/files#r993996726. 

It would essentially only try to add bounds if the coordinate was also a dimension.